### PR TITLE
Fix birthday bugs

### LINKED ISSUE #99: Birthday field allows for invalid dates
Invalid dates such as the 31/02 (31st Feb) are allowed as Birthday, as seen below.

Screenshot 2024-11-01 at 5 10 50 PM

### LINKED ISSUE #100: Birthday field allows for dates beyond current date
Note that this also results in a negative age, which should be resolved by fixing the above

image

### LINKED ISSUE #103: More robust error handling for Birthday
image

Consider having more detailed error messages when handling invalid birthdays, e.g. the date entered as shown is of DD MM YYYY format but it is the month and year that is invalid

### PR DESCRIPTION
**Pull request by AlphaPyke**
Added additional validity barriers to ensure invalid dates such as 31 Feb are not accepted. closes #99 

Added another filter to ensure that dates cannot exceed the current date. closes #100 

Another line added for error message to cover more grounds. closes #103

### CODE REVIEW

File: src/test/java/seedu/address/model/person/BirthdayTest.java [hunk 1]
```diff
@@ -31,10 +31,12 @@ public void isValidBirthday() {
         assertFalse(Birthday.isValidBirthday("phone")); // non-numeric
         assertFalse(Birthday.isValidBirthday("40 10 2991")); // more than 31 days
         assertFalse(Birthday.isValidBirthday("9312 1534")); // not formatted
+        assertFalse(Birthday.isValidBirthday("31 02 2002")); // february should have <= 28 days
 
         // valid birthdays
         assertTrue(Birthday.isValidBirthday("01 01 2002")); // first of month
         assertTrue(Birthday.isValidBirthday("31 10 1908")); // last of month
+        assertTrue(Birthday.isValidBirthday("28 02 2002")); // last of february
```
Comment (meatyturtle) [inline src/test/java/seedu/address/model/person/BirthdayTest.java:39]: nice job testing the boundary values! the course has definitely taught you well

File: src/main/java/seedu/address/model/person/Birthday.java [hunk 2]
```diff
@@ -37,13 +39,22 @@ public Birthday(String birthday) {
...
         }
+
+        try {
+            testDate = LocalDate.of(Integer.parseInt(test.substring(6, 10)), Integer.parseInt(test.substring(3, 5)),
+                    Integer.parseInt(test.substring(0, 2))); // The date is valid
+        } catch (DateTimeException e) {
+            return false; // The date is invalid
+        }
+
+        return !testDate.isAfter(LocalDate.now());
```
Comment (fymbc) [inline src/main/java/seedu/address/model/person/Birthday.java:57]: Good handling of invalid dates

**Review by fymbc — APPROVED** [review]
LGTM.
